alerta/auth/basic_ldap.py

```diff
@@ -2,20 +2,16 @@
 
 import ldap  # pylint: disable=import-error
 from flask import current_app, jsonify, request
-from flask_cors import cross_origin
 
 from alerta.auth.utils import create_token, get_customers
 from alerta.exceptions import ApiError
 from alerta.models.permission import Permission
 from alerta.models.user import User
 from alerta.utils.audit import auth_audit_trail
 
-from . import auth
 
-
-@auth.route('/auth/login', methods=['OPTIONS', 'POST'])
-@cross_origin(supports_credentials=True)
 def login():
```

Comment:
change this line to ...

    def ldap_login():

... and you can revert all the changes you had to do to make testing work.

>At testing level, Python Unit Test only permits a single route for '/auth/login', so I have moved the route to another file called 'main_auth.py' (if you know how to avoid this, let me know because I wasn't able to do it)

Reply:
The change you propose, I did already try before pushing those changes.

The reason I had to do these changes is because if I run the test in an individual way, they run correctly, but If I run them along with the other cases, for some reason the context and the app seems to be already loaded and the LDAP endpoint is not hit.

As you can see in the next executions (with the change you proposed and discarding every change I did), somehow the context is persisted from one test to another
```
TOXENV="postgres tests/test_auth.py::AuthTestCase tests/test_auth_ldap.py::AuthLdapTestCase" make test
```
```
tests/test_auth.py .........
tests/test_auth_ldap.py .<Response streamed [401 UNAUTHORIZED]>
F<Response streamed [401 UNAUTHORIZED]>
F<Response streamed [401 UNAUTHORIZED]>
F<Response streamed [401 UNAUTHORIZED]>
F
```

```
TOXENV="postgres tests/test_auth_ldap.py::AuthLdapTestCase" make test
```
```
tests/test_auth_ldap.py .2020-06-09 10:18:36,011 alerta.app[33381]: [WARNING] LDAP LOGIN request_id=d262d1c9-5bfc-4ef9-bdc2-a7f4698573c4 ip=127.0.0.1
<Response streamed [200 OK]>
.2020-06-09 10:18:36,410 alerta.app[33381]: [WARNING] LDAP LOGIN request_id=0352eae9-57a2-4e5d-b1bc-f741364f2eca ip=127.0.0.1
<Response streamed [200 OK]>
.2020-06-09 10:18:36,795 alerta.app[33381]: [WARNING] LDAP LOGIN request_id=8a292aec-2d7a-4d2d-8fcc-455bd0a2b1f0 ip=127.0.0.1
<Response streamed [401 UNAUTHORIZED]>
.2020-06-09 10:18:37,207 alerta.app[33381]: [WARNING] LDAP LOGIN request_id=8a04a30a-8e3f-4748-bcf1-bf028942344c ip=127.0.0.1
<Response streamed [401 UNAUTHORIZED]>
.
```

The warning you see is a warning I added at the beginning of the login function in basic_ldap.py module and as you can see is only hit when the test case is run by itself.

I've been trying to avoid this case, I searched a way to remove the context to avoid those code changes, but I saw nothing it could help.

If you know a way to avoid that, I would really appreciate if you can tell me

Reply:
I've tried a few things and nothing works so your approach is fine. Can you rename `main_auth.py` to `login.py`. That makes it consistent with `logout.py`.

Also just wondering why you moved `def login()` in `basic.py` to the bottom of the file? If it makes no difference can you move it back? These sorts of changes create lots of unnecessary noise in a PR making it harder to judge what's changed.